### PR TITLE
[Frontend] Make tests toml-schema independent

### PR DESCRIPTION
**Pull request by grwlf**
**Context:** Transition to the quantum device config schema 2

**Description of the Change:** Solve a regarding toml schema 2 udpate in tests by switching our test custom devices from toml text manipulations to the device capability manipulations

**Benefits:** 
* Tests no longer require toml text manipulations.
* Tests now contain simple examples of custom devices.
* toml-specific code is now locates in `catalyst.utils.toml`. 

**Possible Drawbacks:**

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane-lightning/pull/642

### CODE REVIEW

**Comment by dime10** [comment]
Good point, we could just make it compatible with both. I was just going to move to version 2 entirely once the PR is merged. I'm happy with this as well though.

Note there are several other tests that need this update though :)

**Comment by grwlf** [comment]
> Note there are several other tests that need this update though :)

@dime10 , are there? I think I changed the device which is used by several tests, and can't see other places.

**Comment by dime10** [comment]
> > Note there are several other tests that need this update though :)
> 
> @dime10 , are there? I think I changed the device which is used by several tests, and can't see other places.

Have a look at the test_decomposition and test_quantum_control lit tests

**Comment by grwlf** [comment]
> Have a look at the test_decomposition and test_quantum_control lit tests

@dime10 right, thanks! I have make these test toml-independent. It required some further code reshaping, could you please review?

File: frontend/catalyst/utils/toml.py
Comment (dime10) [inline frontend/catalyst/utils/toml.py:65]: I don't see what this has to do with the toml system

File: frontend/catalyst/qfunc.py
Comment (dime10) [inline frontend/catalyst/qfunc.py:76]: What is the reason this code was moved here from the device? I don't think it really fits here and I would like the qfunc patch to be as slim as possible

File: frontend/catalyst/utils/toml.py
Comment (grwlf) [reply frontend/catalyst/utils/toml.py:65]: I moved it here because of the dependency chain: `get_lib_path` called by `get_device_toml_config` called by `get_device_capabilities`. The last function eventually creates `DeviceCapabilities` which is defined in the toml file. Do you have any suggestions?

File: frontend/catalyst/qfunc.py
Comment (grwlf) [reply frontend/catalyst/qfunc.py:76]: The goal is to minimize usage of `config` (the path to a text file). I replaced config loading with capabilities loading and employed it for what we accessed toml-document in a schema-dependant way previously: the device validation and getting the backend info. This way we can write tests devices featuring pre-existing capabilities and still use the same functionality that we use for real devices.

File: frontend/catalyst/utils/toml.py
Comment (dime10) [reply frontend/catalyst/utils/toml.py:65]: The toml files are also on some of those directories, but most are shared library locations we want to link. So I would not move these to the toml module

File: frontend/catalyst/qfunc.py
Comment (dime10) [reply frontend/catalyst/qfunc.py:76]: Could we instead move more code into the device class?

I guess one of the issues is `extract_backend_info` which Erick needed to overwrite in the CUDA support. But there should be a better way to do this rather than put even more code into the qfunc patch.

File: frontend/catalyst/utils/toml.py
Comment (grwlf) [reply frontend/catalyst/utils/toml.py:65]: Keeping these paths in the runtime.py creates a circular dependency. Do you have suggestions on how to solve it? Do you want me to create a separate `utils/paths.py` ? Please let me know

File: frontend/catalyst/qfunc.py
Comment (grwlf) [reply frontend/catalyst/qfunc.py:77]: Done

File: frontend/catalyst/utils/toml.py
Comment (grwlf) [reply frontend/catalyst/utils/toml.py]: After moving code I found that runtime.py contains only a few lines of code, so I renamed it into paths.py.

File: frontend/catalyst/device/qjit_device.py
Comment (lillian542) [inline frontend/catalyst/device/qjit_device.py:167]: Popping in because I was initially going to ask this on #626, but realized that it's actually from here after the PR was rebased to this one. 

Is this an exact copy of the code in `utils.runtime.py`, or are there modifications?

Comment (grwlf) [reply frontend/catalyst/device/qjit_device.py:167]: I don't think I changed the logic of BackendInfo loading, it should be the exact copy.

Comment (lillian542) [reply frontend/catalyst/device/qjit_device.py:167]: Cool - should we either remove it from `utils.runtime.py`? Or import it from there? Is this to get around an import error?

File: frontend/catalyst/qfunc.py
Comment (dime10) [reply frontend/catalyst/qfunc.py:77]: I'm not sure I got my point across, my suggestion was about moving code from utils/runtime into qjit_device, but to move code from here (the QFunc class, which should be as minimal a patch as we can make it) into the device class. For instance, computing device capabilities and verifying them against the Python device is something that would fit much better in the constructor of the `QJITDevice` class, rather that here.

Maybe this is too difficult with the `extract_backend_info` method being located here though. Do you have any ideas about how to remove it?

File: frontend/catalyst/device/qjit_device.py
Comment (grwlf) [reply frontend/catalyst/device/qjit_device.py:167]: Oh, I see. I think something have messed up the pr. It was supposed to remove runtime.py. Will fix

File: frontend/catalyst/qfunc.py
Comment (grwlf) [reply frontend/catalyst/qfunc.py:77]: It is not easy to do. I see it like following:
1. We want to write tests without changing toml text, we want to change device_capabilities directly
2. We need an access to device options so we pass capabilities to the `extract_backend_info`. We might want to override this function (we do this in CUDA).    
3. I only added a single non-trivial line to the qfunc: the one that encodes `program_features`. It does not change the logic of the patch more what I already did by factorizing capabilities and setting `qnode.qjit_device` attribute. I'll write below about the latter.

I see only downsides of moving code from here to device constructor. We still have old api, so by doing this we either duplicate the code, or break old device logic without reason.

> Do you have any ideas about how to remove it?

A better decision might be to move this code into the `qjit` decorator. I set `qnode.qjit_device` because I need it in subsequent PRs: we do need to access the final qjit_device to e.g. check its state. We have all the information required to calculate `qnode.qjit_device` right in the `qjit` when we get the `qnode`. 

But most of all I would like to keep this PR simple. I would suggest to schedule the redesign as a self-contained task rather than adding secondary goals to prs.

File: frontend/catalyst/device/qjit_device.py
Comment (grwlf) [reply frontend/catalyst/device/qjit_device.py:167]: Runtime.py is removed now.

File: frontend/catalyst/qfunc.py
Comment (lillian542) [reply frontend/catalyst/qfunc.py:77]: > We still have old api, so by doing this we either duplicate the code, or break old device logic without reason.

Putting this somewhere where it doesn't fit as well just to accommodate the old device API doesn't seem like the best call to me. Let's not break the old device, certainly, but if we have to copy a few lines of code that will be deleted when we remove the old API, that temporary duplication seems better than having a less optimal organization for the code that we intend to exist long-term. 

I definitely agree that reworking into the `qjit` decorator does sound like a lot of work and out of scope, but adding this to the old device API so we don't need to move things around like this doesn't.

Comment (grwlf) [reply frontend/catalyst/qfunc.py:77]: > I definitely agree that reworking into the `qjit` decorator does sound like a lot of work and out of scope, but adding this to the old device API so we don't need to move things around like this doesn't.

I got your opinion. Let me note that I think that moving the code to device constructors is also out-of scope for this pr.

Comment (grwlf) [reply frontend/catalyst/qfunc.py:77]: OK, so we have named the following options so far:
1. Move the code code to QJITDevice and QJITDeviceNewAPI constructors (Lillian, David?)
2. Move the code to QJITDeviceNewAPI alone (David?)
3. Move the code to the `@qjit` constructor (Sergei)
4. Leave everything as-is (Sergei)

I would prefer the 4th option because I think it has no disadvantages besides adding a line into the qfunc patch. Out of the first two options I would pick the first one because in my opinion it adds less disadvantage to the codebase.
@dime10 are you convinced?

Comment (lillian542) [reply frontend/catalyst/qfunc.py:77]: Makes sense, my only request is if we go with option 4, let's open a story in shortcut to consider option 2 (after removing the old API) or option 3 in the future. Then we can either come back to it or not. 

Does the compilation team have a place to put technical roadmap tasks to consider for future quarters?

Comment (dime10) [reply frontend/catalyst/qfunc.py:77]: Thanks @grwlf, you are right option 4 is reasonable to avoid blocking this PR 👍  
Initializing and verifying device configurations really should be something the device should be doing, so let's add option 1/2 to the backlog. Your suggestion that the original device should be doing it is tricky because 1) we have no control over those devices, so doing it ourselves in the wrapper device seems more practical, and 2) we take the information from the contract (toml) into a form of our choosing (the data-structures you put in place) so there is no reason for third parties to be responsible for that.
The QJIT should also not be concerned with devices, devices are only relevant to quantum sub-routines. QJIT should stay focused on facilitating the jit compilation.

Option 2 outright is not a good idea until the old api is officially replaced by all supported devices.

File: frontend/catalyst/compiler.py
Comment (dime10) [inline frontend/catalyst/compiler.py:36]: `paths` is a bit generic imo (could be dealing with paths as a concept), how about `runtime_environment`?

File: frontend/catalyst/jax_primitives.py
Comment (dime10) [inline frontend/catalyst/jax_primitives.py:79]: It's quite annoying because I've noticed pylint doesn't act the same in all environments. Just ran `pylint frontend` and `abstract-method` comes up when removed, so let's leave it.
```suggestion
# pylint: disable=unused-argument,abstract-method,too-many-lines
```

File: frontend/catalyst/third_party/cuda/primitives/__init__.py
Comment (dime10) [inline frontend/catalyst/third_party/cuda/primitives/__init__.py:27]: I think it's the same here:
```suggestion
# pylint: disable=protected-access,unused-argument,abstract-method,line-too-long
```

File: frontend/catalyst/utils/toml.py
Comment (dime10) [inline frontend/catalyst/utils/toml.py:479]: This module is just here to abstract away interacting with a TOML file right (and abstracting away from schema differences)? Validating device capabilities once the toml file has been processed then seems unrelated to interacting with the toml format. In that case I would prefer having this in the `catalyst/device` submodule, along with some of the other functions that don't interact with the toml file (I'm not 100% where the boundary should be the, probably anything that requires the device instance).

**Review by dime10 — COMMENTED** [review]
Thanks I think this PR is almost good to go. The only thing I would ask to restrict the `utils/toml` module to be more focused on interacting with the toml file format, rather than including functionality related to device capabilities (it is a utility module after all).

File: frontend/catalyst/compiler.py
Comment (grwlf) [reply frontend/catalyst/compiler.py]: Sure, done.

File: frontend/catalyst/utils/toml.py
Comment (grwlf) [reply frontend/catalyst/utils/toml.py]: Done.
I personally think that it is not possible to find a perfect locations for code in Python+OOP style, so I am trying to focus on observable things only, like cyclic dependencies. The problem is in the fact that in this style we define types and methods simultaneously when we define classes. For my projects I (1) don't use "normal" classes (2) create `types.py` for types and dataclases (3) group functions in -py files according to their semantics.

File: frontend/catalyst/qfunc.py
Comment (grwlf) [reply frontend/catalyst/qfunc.py]: Ok, done.

@lillian542 I have added a todo-comment [here](https://github.com/PennyLaneAI/catalyst/blob/2bb3ede8507574ec026b178421513035a95d0c58/frontend/catalyst/qfunc.py#L71)

File: frontend/catalyst/utils/toml.py
Comment (grwlf) [reply frontend/catalyst/utils/toml.py]: Later renamed to runtime_environment as suggested

File: doc/changelog.md
Comment (dime10) [inline doc/changelog.md:101]: This change is not obversable for the user in the distribution, so doesn't need a changelog entry. Typically we exclude changelog entries for test-only, ci-only, doc-only, and repo-only changes.

File: frontend/catalyst/utils/toml.py
Comment (dime10) [reply frontend/catalyst/utils/toml.py]: What about get_device_toml_config and get_device_capabilities, their code seems to be operating primarily on the device, and the calls into toml functions like read_toml_file and load_device_capabilities?

**Review by dime10 — APPROVED** [review]

File: doc/changelog.md
Comment (grwlf) [reply doc/changelog.md:101]: ok done

Comment (dime10) [reply doc/changelog.md:101]: I don't think the commit changed anything?

File: frontend/catalyst/utils/toml.py
Comment (grwlf) [reply frontend/catalyst/utils/toml.py]: Done

File: frontend/catalyst/compiler.py
Comment (grwlf) [reply frontend/catalyst/compiler.py]: Sorry, done in https://github.com/PennyLaneAI/catalyst/pull/626/files

File: frontend/catalyst/utils/toml.py
Comment (dime10) [reply frontend/catalyst/utils/toml.py]: Where?

**Pull request merged**
